exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/metadata/MetadataControllerBatch.java

```diff
@@ -127,113 +126,93 @@ protected MetadataControllerBatch(MetadataControllerPOP popConfig,
         ? null
         : popConfig.getContext().metadataToHandle().stream()
             .collect(Collectors.toMap(MetadataInfo::identifier, Function.identity()));
-    this.metadataUnits = new ArrayList<>();
-    this.statisticsCollector = new StatisticsCollectorImpl();
     this.columnNamesOptions = new ColumnNamesOptions(context.getOptions());
   }
 
-  protected boolean setupNewSchema() {
-    container.clear();
-    container.addOrGet(MetastoreAnalyzeConstants.OK_FIELD_NAME, Types.required(TypeProtos.MinorType.BIT), null);
-    container.addOrGet(MetastoreAnalyzeConstants.SUMMARY_FIELD_NAME, Types.required(TypeProtos.MinorType.VARCHAR), null);
-    container.buildSchema(BatchSchema.SelectionVectorMode.NONE);
-    container.setEmpty();
-    return true;
-  }
-
   @Override
   public IterOutcome innerNext() {
-    IterOutcome outcome;
-    boolean finishedLeft;
-    if (finished) {
-      return IterOutcome.NONE;
-    }
+    while (state != State.FINISHED) {
+      switch (state) {
+        case RIGHT: {
 
-    if (!finishedRight) {
-      outcome = handleRightIncoming();
-      if (outcome != null) {
-        return outcome;
+          // Can only return NOT_YET
+          IterOutcome outcome = handleRightIncoming();
+          if (outcome != null) {
+            return outcome;
+          }
+          break;
+        }
+        case LEFT: {
+
+          // Can only return NOT_YET
+          IterOutcome outcome = handleLeftIncoming();
+          if (outcome != null) {
+            return outcome;
+          }
+          break;
+        }
+        case WRITE:
+          writeToMetastore();
+          createSummary();
+          state = State.FINISHED;
+          return IterOutcome.OK_NEW_SCHEMA;
+
+        case FINISHED:
+          break;
+
+        default:
+          throw new IllegalStateException(state.name());
       }
     }
+    return IterOutcome.NONE;
+  }
 
+  private IterOutcome handleRightIncoming() {
     outer:
-    while (true) {
-      outcome = next(0, left);
+    for (;;) {
```

Comment:
Are there any reasons for replacing `while (true)` with `for (;;)`?

Reply:
@vvysotskyi, thanks for the review. The `for (;;)` is a bit more terse; happy to put it back if doing so is clearer.

Reply:
Yes, please replace it if possible to avoid IDE warnings.

Reply:
Interesting. Which IDE warns? With which warning? Eclipse is fine with either syntax. I can perhaps add a warning in Eclipse to match your IDE.

Reply:
I use IntelliJ IDEA, the warning was the following: `'for' loop may be replaced with 'while' loop`.
Its description is
```
Reports for loops which contain neither initialization or update components, and can thus be replaced by simpler while statements. Example:
  for(; exitCondition(); ) {
    process();
  }
This loop can be replaced with
  while(exitCondition()) {
    process();
  }
A fix action is also available for other for loops, so you can replace any for loop with while. 
Use the checkbox below if you wish this inspection to ignore for loops with trivial or non-existent conditions.

```

Reply:
Thanks for the explanation! I've updated DRILL-7352, our roll-up of coding standards, with the `while (true)` preference.

Reply:
Thanks for adding it to the Jira! Some time ago I looked into the way how we can apply separate checkstyle rules without enforcing other rules we don't want to touch, but tools like https://github.com/diffplug/spotless or IDEA autoformatting updates the code using a complete list of rules.